Revamped and enhanced webapp protocol in I-D and spec.yaml (#368)
cs3org/OCM-API: IETF-RFC.md

    support for acting as a Receiving Server.

    Fields:
    - webdav (string) - The top-level WebDAV [RFC4918] path at this
      endpoint.  In order to access a Remote Resource, implementations
      SHOULD use this path as a prefix (see sharing examples).
    - webdav-receive (object) - Advertised by implementations that
      support receiving WebDAV shares.  It contains a `uri` property
      whose value MUST be either `"absolute"` or `"relative"`,
      signalling the URI format this endpoint accepts.  Note that
      older implementations MAY not support this property.
    - webapp (object) - Advertised, as an empty object, by
      implementations that support sending WebApp shares.
    - webapp-receive (object) - Advertised by implementations that
      support receiving WebApp shares.  It contains a `targets`
      array listing the ways this endpoint is able to present a
      WebApp share to the user.  A subset of:
      - `blank` - the endpoint can open the URI in a new window or
        tab.
      - `redirect` - the endpoint can navigate the browser to the
        URI, replacing the current page.
      - `iframe` - the endpoint can embed the URI in an iframe

>         within its own UI, and can set CORS headers.

Reviewer: Same iframe wording as in the OpenAPI file: I think this should talk about the sender allowing framing, not the receiver setting CORS headers.

```suggestion
      - `iframe` - the endpoint can embed the URI in an iframe
        within its own UI when the Sending Server allows framing by
        this receiver.
```

    - `share` allows re-share rights on the Resource.  This only
      applies to web apps that provide a mechanism for re-sharing.
  - REQUIRED requirements (array of strings) -
    The requirements that the sharee MUST fulfill to
    access the Resource.  The requirements MUST at least include
    `must-exchange-token`.  If multiple protocols are present in the
    share payload, the requirements for the different protocols MUST
    agree.  For example, if a webapp share is sent in the same payload
    as a webdav share, both protocols MUST carry the same
    requirements, and both requirement arrays MUST include
    `must-exchange-token`.
  - OPTIONAL sharedSecret (string)
    A secret for accessing the remote web app.  To give access to the
    remote app, the receiver MUST first exchange this value at the
    Sending Server's {tokenEndPoint} using the Code Flow, then perform
    an HTTP POST request to the given `uri` with the resulting bearer
    token in a form field named `access_token` (see
    [Resource Access](#resource-access)).  The shared secret MUST NOT
    be exposed to the browser and MUST NOT appear in any URI.  In a

>     multi-protocol Share that also offers WebDAV, the access

Reviewer 2: Maybe we can now drop this sentence, the above part is stronger.

    especially in case of `datatx` access type.
* Protocol details for `webapp` MAY contain:
  - REQUIRED uri (string)
    A URI to a client-browsable view of the Shared Resource, such
    that users MAY use a web application available at the Sending
    Server.  The URI MUST be absolute, including a hostname.  In
    case the underlying Resource is a folder, the URI MUST act as a
    root path, such that files located within the folder are made
    accessible in the web app by appending their relative path to
    the URI.
  - REQUIRED targets (array of strings) - How the recipient SHOULD

>     present the URI to the user.  MUST NOT be empty.  A subset of:

Reviewer: Since `targets` is now required, I think the later `defaulting to blank` wording becomes a bit confusing.

Could we make the non-empty requirement explicit here and handle missing/empty targets as invalid?

```suggestion
  - REQUIRED targets (array of strings) - How the recipient SHOULD
    present the URI to the user.  The `targets` array MUST NOT be
    empty.  A subset of:
```

    - `blank` signals the recipient to open the URI in a new window
      or tab.
    - `redirect` signals the recipient to navigate the browser to
      the URI, replacing the current page.
    - `iframe` signals the recipient to embed the URI in an iframe

>       within its own UI.  CORS headers MUST be properly set.

Reviewer: Same iframe point here too: I think this should use framing-policy language instead of CORS.

```suggestion
    - `iframe` signals the recipient to embed the URI in an iframe
      within its own UI when the Sending Server allows framing by this
      receiver.
```

    A Sending Server MUST NOT offer a target that the recipient did
    not advertise in its `webapp-receive` discovery property.
  - REQUIRED permissions (array of strings) -
    The permissions granted to the sharee.  MUST NOT be empty.
    A subset of:
    - `view` allows access to the web app in view-only mode.
    - `read` allows read and download access via the web app.
    - `write` allows full editing rights via the web app.
    - `share` allows re-share rights on the Resource.  This only
      applies to web apps that provide a mechanism for re-sharing.
  - OPTIONAL sharedSecret (string)
    A secret for accessing the remote web app, such as a bearer
    token.  To give access to the remote app, the receiver MUST
    perform an HTTP POST request to the given `uri`, passing the
    shared secret in a form field named `access_token` (see
    [Resource Access](#resource-access)).  To prevent leaking it in
    logs it MUST NOT appear in any URI.  In a multi-protocol Share
    that also offers WebDAV, the access requirements provided in the
    `webdav` part (such as `must-exchange-token`) MUST apply to

>     `webapp` accesses as well.

Reviewer: Similar to the OpenAPI comment: I think the safer contract is that `sharedSecret` is exchanged server-to-server first, and only the resulting bearer token is posted to the WebApp.

If direct POST of the raw `sharedSecret` is intentional, then I think it needs a bit more security text around lifetime/replay/logging.

```suggestion
    A secret for accessing the remote web app.  To give access to the
    remote app, the receiver MUST first exchange this value at the
    Sending Server's {tokenEndPoint} using the Code Flow, then perform
    an HTTP POST request to the given `uri` with the resulting bearer
    token in a form field named `access_token`.  The shared secret MUST
    NOT be exposed to the browser and MUST NOT appear in any URI.  In a
    multi-protocol Share that also offers WebDAV, the access
    requirements provided in the `webdav` part (such as
    `must-exchange-token`) MUST apply to `webapp` accesses as well.
```

Reviewer: I feel like this request is bit of a nitpick since I have not read the full webapp implementation you did in Nextcloud, so I might have completely got it wrong.

Author: No, indeed, this is how it is being implemented in Nextcloud, and I think you reading is correct here.

  - OPTIONAL appName (string)
    A human-friendly name of the web application, to be used in user
    interfaces when referring to this Share.
  - OPTIONAL appIcon (string)
    A URI to an icon representing the web application, to be used in
    user interfaces when referring to this Share.  An embedded data
    URI MAY be used; alternatively, if a regular URI is used, it MUST

>     be absolute, including a hostname.

Reviewer: This may be worth constraining a little because receiver UIs will likely render this value directly.

Could we say that data URIs must still be image resources and should be rendered only in an inert image context?

```suggestion
    A URI to an icon representing the web application, to be used in
    user interfaces when referring to this Share.  An embedded data
    URI MAY be used if it identifies an image resource; alternatively,
    if a regular URI is used, it MUST be absolute, including a
    hostname.  Receiving Servers MUST render the icon only in an inert
    image context and MAY reject unsupported or unsafe image types.
```

   identified (e.g. because `protocol.options` is undefined), then
   the receiver SHOULD discard the share as invalid.
6. For the specific case where `protocol.webapp` is available and the
   receiver wants to use it, the receiver MUST present the web app to
   the user by opening `protocol.webapp.uri` using one of the targets
   listed in `protocol.webapp.targets` (defaulting to `blank`), chosen
   among those the receiver advertises in its `webapp-receive`
   discovery property.  If a `protocol.webapp.sharedSecret` is present,
   the receiver MUST NOT place it in the URI; instead it MUST deliver
   it to the web app via an HTTP POST to `protocol.webapp.uri` with the
   secret carried in a form field named `access_token`.  This is
   typically achieved with an auto-submitting HTML form whose `target`
   attribute selects the chosen presentation (e.g. an iframe name,
   `_blank`, or `_top`).  When the Share also requires token exchange
   (see step 3.1, applied via the `webdav` part of a multi-protocol
   Share), the receiver MUST first exchange the `sharedSecret` at the
   Sending Server's {tokenEndPoint} and POST the resulting bearer token

>    as the `access_token` value.

Reviewer: Could we avoid the default here and define the no-common-target case?

That would make the receive-side behavior deterministic when the sender's `targets` and the receiver's `webapp-receive.targets` do not overlap. This also keeps the raw `sharedSecret` out of the browser.

```suggestion
6. For the specific case where `protocol.webapp` is available and the
   receiver wants to use it, the receiver MUST present the web app to
   the user by opening `protocol.webapp.uri` using a target selected
   from the intersection of `protocol.webapp.targets` and the targets
   advertised in the receiver's `webapp-receive` discovery property.
   If this intersection is empty, the receiver MUST treat the `webapp`
   option as unusable for this Share.  If a
   `protocol.webapp.sharedSecret` is present, the receiver MUST NOT
   place it in the URI and MUST NOT expose it to the browser.  Instead,
   the receiver MUST first exchange it at the Sending Server's
   {tokenEndPoint} using the Code Flow, then deliver the resulting
   bearer token to the web app via an HTTP POST to
   `protocol.webapp.uri` with the token carried in a form field named
   `access_token`.  This is typically achieved with an auto-submitting
   HTML form whose `target` attribute selects the chosen presentation
   (e.g. an iframe name, `_blank`, or `_top`).
```

There are several areas that are not covered by this specification.
Most importantly we do not provide a way of establishing trust between
servers, even though some features of the protocol rely on trust, such

> as the `mfa-enforced` requirement.

Reviewer 2: This was addressed by https://github.com/cs3org/OCM-API/pull/366, can we review/merge that one first?
